### ISSUE TITLE
Oversampling without smoothing

### ISSUE THREAD

**Issue by HybridDog** [minetest/minetest#6976]
I think the algorithm I tried for mipmapping (#6917) works great for oversampling.
Please view the screenshots in 1:1 resolution or with nearest neighbour upscaling (integer or big scaling factor).

nearest neighbour downscaled (corresponds most to no oversampling):
![nn](https://user-images.githubusercontent.com/3192173/35473747-d02a57b4-0384-11e8-8e27-2e3531bcd1b2.png)
![tulip_nn](https://user-images.githubusercontent.com/3192173/35473829-d461dd6a-0385-11e8-9b0b-91a7c64f8ce9.png)

ssim based downscaled:
![ssim](https://user-images.githubusercontent.com/3192173/35473772-3292e9de-0385-11e8-9fa8-01dda881ba31.png)
![tulip_ssim](https://user-images.githubusercontent.com/3192173/35473827-d0415170-0385-11e8-9c79-fef4d7b3fda5.png)

As you can see, oversampling with this downscaling algorithm
* increases details, see the tulip,
* rids Moire-effects (see the bricks),
if mipmapping is used, there're more details instead of Moire effect reduction, 
* has an anti-aliasing effect,
* does not smooth the image (or insignificantly), this would happen with linear or bicubic downscaling

l think, it could be implemented with parallelization (shaders), the speed of the algorithm lies between linear and bicubic downscaling.
Also, if implemented on the gpu it maybe doesn't have a noticeable performance impact because minetest is cpu-intensive.

**Comment by Calinou** [comment]
Remember that even 2× SSAA is very expensive, as it requires rendering four times the original amount of pixels. For example, to output to a 1920×1080 window, you need to render in 3840×2160. Since most people with high-end graphics cards have displays above 1920×1080, it's going to be even worse for them.

**Comment by HybridDog** [comment]
You're right, I think there're also graphic chips which don't support rendering at very high resolution.
Still, if you have a fast graphics card, the cpu directs the amount of lag in minetest (other than in sauerbraten).

**Comment by Calinou** [comment]
> You're right, I think there're also graphic chips which don't support rendering at very high resolution.

Those graphics cards would be too slow to render high resolutions at playable framerates anyway.

**Comment by nerzhul** [comment]
please respect the new issue template...

**Comment by HybridDog** [comment]
ok.

**Comment by paramat** [comment]
Some of the template sections will be irrelevant for some issues. Version / OS / hardware is mostly for bugs.

**Comment by HybridDog** [comment]
> Some of the template sections will be irrelevant for some issues.

l think so, too.

**Comment by paramat** [comment]
1 year bump. Any core dev support?

**Comment by HybridDog** [comment]
I thought about adding a configurable threshold resolution, oversampling is automatically en-/disabled when changing the window resolution.

**Comment by paramat** [comment]
Issue seems useful to keep open for consideration.

**Comment by Calinou** [comment]
I still like the idea of enabling SSAA manually with a setting like `oversampling = 2` in  `minetest.conf`, as recent rendering optimizations make this more viable on high-end machines.

**Comment by devshgraphicsprogramming** [comment]
Properly configured Anisotropic Filtering is free

**Comment by Calinou** [comment]
@devshgraphicsprogramming Anisotropic filtering doesn't solve this issue fully when using nearest-neighbor interpolated textures :slightly_smiling_face:

**Comment by devshgraphicsprogramming** [comment]
OpenGL has two texture filtering modes that can be set separately (which irrlicht doesn't expose):
- Minification (when you go away from the texture)
- Magnification

you could set mag filter to NEARST, and min to LINEAR_MIPMAP_LINEAR

Also Irrlicht SVN maybe has direct mip-map control, what @HybridDog is proposing is simply generating all the mip-maps properly (offline) with nice filters and uploading them explicitly instead of relying on OpenGL mipmap generation hints to do the work for you from the base level.

There's lots of tools (GIMP, Nvidiai Texture Tools, AMD Compressonator) that willl generate mip-maps of textures for you with nice filters like Lanczos or Kalman (or Kelemen, don't remember how its spelt)


MSAA has nothing to do with this, also MSAA is an arcane technique that belongs to the 2000s

**Comment by HybridDog** [comment]
> OpenGL has two texture filtering modes that can be set separately […]

Does Minetest use LINEAR_MIPMAP_NEAREST when mipmapping is enabled (and the magnification filter is NEAREST)?

> […] what @HybridDog is proposing is simply generating all the mip-maps properly (offline) […]

That's another issue: #6917
I've opened this issue separately for oversampling.

**Comment by HybridDog** [comment]
I have implemented the oversampling with shaders in an opengl example program:
[ssim_based_realtime_snippets_20191225.zip](https://github.com/minetest/minetest/files/4434252/ssim_based_realtime_snippets_20191225.zip)

I don't know how to put this into minetest, which uses Irrlicht. The code changes could involve rendering in a higher resolution (minetest already has an undersampling setting, where it renders in lower resolution), and applying the downscaling shaders (this should be similar to adding post-processing).

**Comment by HybridDog** [comment]
I have [implemented the algorithm in Minetest](https://github.com/HybridDog/minetest/tree/m_oversampling). I hope that I haven't done severe mistakes.
When I apply the algorithm for all channels and not only luma, the colour looks wrong. I'm not sure why this happens, maybe because of numerical errors. When I do it with an external program, which has more than 16 bit floating point precision, the result looks better.
Here are some pictures:

### Linear
![linear](https://user-images.githubusercontent.com/3192173/86518762-22399f80-be34-11ea-9f52-f303467c2f79.png)
![excerpt_linear](https://user-images.githubusercontent.com/3192173/86518806-8b211780-be34-11ea-9d2e-fb5e1e686d7e.png)

### SSIM-based, luma only
![ssim_luma](https://user-images.githubusercontent.com/3192173/86518766-26fe5380-be34-11ea-998d-d39c6186b172.png)
![excerpt_ssim_luma](https://user-images.githubusercontent.com/3192173/86518809-8eb49e80-be34-11ea-9146-46a982ce6359.png)

### No oversampling (default minetest):
![disabled](https://user-images.githubusercontent.com/3192173/86518802-793f7480-be34-11ea-91aa-e7786655e395.png)
![excerpt_disabled](https://user-images.githubusercontent.com/3192173/86518805-8a888100-be34-11ea-9c76-f7f64ed0fddf.png)

### SSIM-based, y, cb and cr
![ssim_all](https://user-images.githubusercontent.com/3192173/86518765-25349000-be34-11ea-8f2d-1a2eda5a1ed6.png)
![excerpt_ssim_all](https://user-images.githubusercontent.com/3192173/86518807-8bb9ae00-be34-11ea-8f5a-49846299a1f8.png)

### No oversampling version downscaled SSIM-based, y, cb and cr with an external program
![tmp_screenshot_smaller](https://user-images.githubusercontent.com/3192173/86518907-8b6de280-be35-11ea-9df3-636b3e8a072c.png)

**Comment by HybridDog** [comment]
After some changes the y, cb and cr variant works now. Apparently it is possible to manually set the sharpening for the pixels with precision problems.
### linear SSAA
![ex_lin](https://user-images.githubusercontent.com/3192173/86519539-ef93a500-be3b-11ea-959d-c2f1a232790f.png)
### disabled
![ex_off](https://user-images.githubusercontent.com/3192173/86519540-f0c4d200-be3b-11ea-830d-d70e2d946c46.png)
### SSIM-based SSAA
![ex_ssim](https://user-images.githubusercontent.com/3192173/86519541-f0c4d200-be3b-11ea-8619-ce3bf2d93433.png)
![2020-07-04-211141_1091x645_scrot](https://user-images.githubusercontent.com/3192173/86519557-194ccc00-be3c-11ea-8ef6-2900838a6ca7.png)

**Comment by grapereader** [comment]
Interestingly, your changes to RenderingCorePlain are all that is necessary to add in an FXAA shader. https://github.com/grapereader/minetest/commit/7ba37fefa7a7de646f3a7a36943a5ff0e95d8a0d

So that's fun 😅

Would be awesome if Minetest supported screen space post-processing more cleanly so it didn't have to be hacked in like this. But I digress.

**Comment by Calinou** [comment]
@grapereader I agree it would be nice to offer a built-in FXAA option, since it can look quite good at 1440p or 4K. Feel free to open a pull request for this :slightly_smiling_face:

**Comment by HybridDog** [comment]
There's an abandoned Pull Request for post processing: #9895

**Comment by grapereader** [comment]
I've been playing around some more and honestly, simple SSAA (as implemented here) works great at 1080p and should be an option in the base game. Since Minetest rendering is already so fast, I easily get 60fps despite the double size render. And it just looks _beautiful_.

On the other hand people with 4K would probably want something faster like FXAA. That's also doable.

I spent all night trying to get SMAA to work hoping that would be the one-size-fits-all solution but it's been a crap shoot. Pipeline works great up to and including edge detection but I can't get the damned thing to calculate weights properly. It's just too complex to grapple with. Oh well.

Tested some other single-pass algorithms like DLAA, DDAA, NFAA. They work fine but FXAA is still king here. DDAA is a close second.

Still, we need a generalized post-processing pipeline before any of this could be merged. I checked out https://github.com/minetest/minetest/pull/9895 (thanks for the link). It is on the right path. It adds a new RenderingCore, which is exactly how I would have done it. I don't know if the complaints against the additional 3D mode (which seem to have been the main blocker) are tenable without either:

- a) Adding a new rendering abstraction that executes after a RenderingCore (which seems like overkill just to preserve esoteric modes like anaglyph, and might actually break some of the more advanced post-processing requiring things like depth buffer, which the existing RenderingCores butcher to high heaven)

- b) Somehow re-working these modes to be extensions that can be passed -into- our new RenderingCore. Again, is this really worth it?
- c) ???

At this point I'm not sure if I should be commenting here or on that PR. At least it is linked here, now, because it was mentioned.

**Comment by grapereader** [comment]
Some screenshots just for fun.

Linear SSAA
![linear-ssaa](https://user-images.githubusercontent.com/1329614/113644067-b1304e00-9638-11eb-9226-7a9338711941.png)

FXAA
![fxaa](https://user-images.githubusercontent.com/1329614/113644383-7f6bb700-9639-11eb-96a6-54c582d12ed0.png)

Phenomenal. MSAA's worst nightmare.

**Comment by HybridDog** [comment]
I have reimplemented SSAA with SSIM-based perceptual downscaling for the current Minetest version: https://github.com/HybridDog/minetest/tree/m_oversampling_new.
I don't know if it's implemented correctly of if I have made mistakes.
The images should be viewed in 1:1 resolution without scaling.

By the way: I noticed that I've accidentally swapped the EOTF and OETF in the colour conversions before in my old implementation, so I've hidden the screenshots above now.

## No anti-aliasing
![no_aa](https://github.com/minetest/minetest/assets/3192173/5056cd2b-28e5-42b9-9ca9-0105be1d648d)

## SSAA currently available in Minetest
This image looks darker because downscaling is done without gamma correction. 
![ssaa_mt_default](https://github.com/minetest/minetest/assets/3192173/51060187-a612-4ed4-b2d6-bd3c7950b3eb)

## SSAA with SSIM-based perceptual downscaling

### All YCbCr channels sharpened
![ssaa_ssim](https://github.com/minetest/minetest/assets/3192173/5410acfa-728f-46e2-8403-831ab12e2c3c)

### Only the Y channel sharpened
![ssaa_ssim_y_only](https://github.com/minetest/minetest/assets/3192173/0c7b9f70-5561-4e70-b617-915e0c570056)

### No sharpening
This should be the same as SSAA currently available in Minetest except that it has gamma correction. 
![ssaa_no_sharpening](https://github.com/minetest/minetest/assets/3192173/a0681384-1edb-4437-bce6-b792aa900393)

**Comment by HybridDog** [comment]
I've experimented with different colour spaces, different values for the numerical precision problem prevention, and tried mpv's shader.
For the experiments I've done small adjustments to [my implementation in C](https://gist.github.com/HybridDog/dd95a99d411972f030fb18543280ad60).
The full images are here: [ssim_based_downscaling_experiments.zip](https://github.com/minetest/minetest/files/13259704/ssim_based_downscaling_experiments.zip).
Below are my observations regarding colour spaces and the numerical precision problem prevention.

# Effect of the chosen colour space

If I use linearised sRGB instead of linear YCbCr, visible colour fringes appear, such as blue pixels in the bricks and red pixels in the text which was originally just white on bright blue background.
If I do the sharpening only on the Y channel, there is less introduced colour contrast and lower memory and computation requirement. I don't know if this looks better or worse than doing it on all YCbCr channels.

Downscaled text; image order: RGB, YCbCr, YCbCr with only Y sharpened
![text_rgb](https://github.com/minetest/minetest/assets/3192173/ad6002e2-0fce-4ca3-baf5-56462fc05c3c)![text_ycbcr](https://github.com/minetest/minetest/assets/3192173/b8ccfac3-608b-403a-be27-575bc197e53e)![text_ycbcr_y_only_sharpened](https://github.com/minetest/minetest/assets/3192173/a2fec0b0-af3d-42df-bf00-52f92d83dd06)

Downscaled bricks; image order: RGB, YCbCr, YCbCr with only Y sharpened
![brick_side_rgb](https://github.com/minetest/minetest/assets/3192173/fcd79a93-e2fa-4a3e-8c58-c313c3846a27)![brick_side_ycbcr](https://github.com/minetest/minetest/assets/3192173/abe04a56-7259-4218-878b-e1568c3ea14a)![brick_side_ycbcr_y_only_sharpened](https://github.com/minetest/minetest/assets/3192173/68f71740-1f23-4fec-9775-c824c2063310)

# Effect of the threshold to prevent numerical errors

[There is a threshold](https://gist.github.com/HybridDog/dd95a99d411972f030fb18543280ad60#file-ssim_perceptual_downscaling-c-L245) which is required to prevent numerical errors. If the image is stored with 16 bit floats instead of 32 bit floats, I need to increase this threshold to 0.0005 to avoid artifacts, so a lower threshold requires more memory to store images.
In my implementation, if the threshold is not exceeded at a pixel, there is a fixed sharpening (`r_all[i] = 2.0f;`); I don't know if sharpening in this case is good or bad behaviour.
With a 0.0005 threshold, the `i` and `n` melt together in `Minetest` in the text example and there is a loss of detail in the bricks example. With a 0.05 threshold, the algorithm no longer gives good results, degenerating to image sharpening (in my implementation).
If the downscaling algorithm is used for SSAA, perhaps it's better to use 32 bit floats and sharpen only the Y channel with a lower threshold than to use 16 bit floats and sharpen all YCbCr channels, or perhaps it's not.

Downscaled text with different threshold values; image order (thresholds): 0.000001, 0.0005, 0.05:
![text_a_0 000001](https://github.com/minetest/minetest/assets/3192173/f16dad4b-b660-44fb-829b-ec7cf160436b)![text_b_0 0005](https://github.com/minetest/minetest/assets/3192173/71c70d66-9e44-4e42-a5ad-20be18b2c0f6)![text_c_0 05](https://github.com/minetest/minetest/assets/3192173/262bade9-c42b-4890-8ce0-7c6dda5998bb)

Downscaled bricks with different threshold values; image order (thresholds): 0.000001, 0.0005, 0.05:
![bricks_a_0 000001](https://github.com/minetest/minetest/assets/3192173/971b600e-a502-4727-b0e4-7df59acf8377)![bricks_b_0 0005](https://github.com/minetest/minetest/assets/3192173/0e3592cb-f016-49c3-8298-b8d5c672c9e5)![bricks_c_0 05](https://github.com/minetest/minetest/assets/3192173/209247e7-9c9d-4851-8036-dd0d75ecf0c7)

**Comment by HybridDog** [comment]
> If the downscaling algorithm is used for SSAA, perhaps it's better to use 32 bit floats and sharpen only the Y channel with a lower threshold than to use 16 bit floats and sharpen all YCbCr channels, or perhaps it's not.

I've changed my implementation to use 0.000001 and sharpen only the Y channel and I think it looks better.
